compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala

```diff
@@ -682,16 +682,304 @@ abstract class GenJSCode extends plugins.PluginComponent
             "repeated parameters.")
       }
 
-      // Implementation restriction
-      for (tree <- secondaryCtorTrees) {
-        reporter.error(tree.pos,
-            "Implementation restriction: Scala.js-defined JS classes cannot " +
-            "have secondary constructors")
+      withNewLocalNameScope {
+        val infoBuilder = new MethodInfoBuilder()
+        infoBuilder.setEncodedName("constructor")
+        infoBuilder.setIsExported(true)
+        infoBuilder.setIsStatic(false)
+
+        val infoBuilderOpt = Some(infoBuilder)
+
+        val primaryCtor =
+          genMethodWithInfoBuilder(primaryCtorTree, infoBuilderOpt).get._1
+        val secondaryCtors = secondaryCtorTrees.map { tree =>
+          genMethodWithInfoBuilder(tree, infoBuilderOpt).get._1
+        }
```

Comment:
This probably means we should defer the `partition` above until later, so we can generically handle the primary constructor and the secondary constructors at this point.